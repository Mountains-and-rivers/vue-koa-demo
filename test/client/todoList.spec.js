import Vue from 'vue'
import ElementUI from 'element-ui'
import { mount } from '@vue/test-utils'
import TodoList from '@/components/TodoList'
import flushPromise from 'flush-promises'

Vue.use(ElementUI)
jest.mock('axios', () => ({
  get: jest.fn()
    // for test1
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true,
        msg: 'breezymelon'
      }
    }))
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true,
        todoTotal: 1,
        todoList: [
          {
            todoDetail: 'Todo Detail',
            todoState: 'todo',
            todoTime: '1544085337102'
          }
        ]
      }
    }))
    // for test2
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true,
        msg: 'breezymelon'
      }
    }))
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true,
        todoTotal: 0
      }
    }))
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true,
        todoTotal: 1,
        todoList: [
          {
            todoDetail: 'A new todo',
            todoState: 'todo',
            todoTime: '1544085337102'
          }
        ]
      }
    }))
    // for test3
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true
      }
    }))
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true,
        todoTotal: 1,
        todoList: [
          {
            todoDetail: 'A new todo',
            todoState: 'todo',
            todoTime: '1544085337102'
          }
        ]
      }
    }))
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true,
        todoTotal: 1,
        todoList: [
          {
            todoDetail: 'A new todo',
            todoState: 'done',
            todoTime: '1544085337102'
          }
        ]
      }
    }))
    // for test4
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true
      }
    }))
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true,
        todoList: [
          {
            todoState: 'todo'
          },
          {
            todoState: 'done'
          }
        ]
      }
    }))
    // for test5
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true
      }
    }))
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true,
        todoTotal: 1,
        todoList: [
          {
            todoState: 'todo',
            todoDetail: 'A new detail'
          }
        ]
      }
    }))
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true,
        todoTotal: 1,
        todoList: [
          {
            todoState: 'todo',
            todoDetail: 'An updated detail'
          }
        ]
      }
    }))
    // for test6
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true
      }
    }))
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true
      }
    })),
  post: jest.fn()
    // for test2
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true
      }
    }))
    // for test3
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true
      }
    }))
    // for test5
    .mockImplementationOnce(() => Promise.resolve({
      data: {
        success: true
      }
    }))
}))

const { getComputedStyle } = window
window.getComputedStyle = function getComputedStyleStub (el) {
  return {
    ...getComputedStyle(el),
    transitionDelay: '',
    transitionDuration: '',
    animationDelay: '',
    animationDuration: '',
    getPropertyValue: jest.fn()
  }
}

describe('TodoList.vue', () => {
  let wrapper
  beforeEach(() => {
    wrapper = mount(TodoList, {
      sync: false
    })
  })

  test('should get correct todolist', async () => {
    sessionStorage.__STORE__.hasLogin = true

    expect(wrapper.vm.todoTotal).toBe(1)
    expect(wrapper.vm.todoList).toHaveLength(1)
  })

  test('should add new todo when press the Enter + Ctrl key', async () => {
    wrapper.setData({ todoToAdd: 'A new todo' })
    expect(wrapper.vm.todoTotal).toBe(0)
    await wrapper.vm.submitTodo()
    await flushPromise()

    expect(wrapper.vm.todoTotal).toBe(1)
  })

  test('should toggle todo todoState when click todoState button', async () => {
    expect(wrapper.vm.todoList.filter(todo => todo.todoState === 'todo')).toHaveLength(1)
    await wrapper.vm.completeTodo({ _id: 1 })

    expect(wrapper.vm.todoList.filter(todo => todo.todoState === 'todo')).toHaveLength(0)
  })

  test('todo input should exist when click todo item', async () => {
    wrapper.find('ul li:nth-child(1) .item-container p').trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.todoList[0].inputShow).toBeTruthy()
    wrapper.find('ul li:nth-child(2) .item-container p').trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.todoList[1].inputShow).toBeFalsy()
  })

  test('todo should be updated when todo item input emit blur event', async () => {
    wrapper.vm.todoList[0].inputShow = true
    wrapper.find('ul li:first-child div p input').trigger('blur')
    await flushPromise()
    expect(wrapper.vm.todoList[0].todoDetail).toBe('An updated detail')
  })

  test('sessionStorage should be changed when component is about to be destroyed', async () => {
    wrapper.setData({ page: 2 })
    expect(sessionStorage.__STORE__.pagination).toBeFalsy()
    wrapper.destroy()
    expect(Number(sessionStorage.__STORE__.pagination)).toBe(2)
  })

  test('todoList.vue snapshot test', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
