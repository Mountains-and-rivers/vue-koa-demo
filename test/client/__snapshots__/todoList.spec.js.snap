// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList.vue todoList.vue snapshot test 1`] = `
<div
  class="todo-container"
>
  <div
    class="todo-input el-textarea"
  >
    <textarea
      autofocus="autofocus"
      class="el-textarea__inner"
      placeholder="请输入待办事项，Ctrl + Enter 添加"
      rows="2"
      style="resize: none; min-height: NaNpx;"
    />
  </div>
   
  <div
    class="state-btns"
  >
    <div
      class="el-radio-group"
      role="radiogroup"
    >
      <label
        aria-checked="true"
        class="el-radio-button el-radio-button--small is-active"
        role="radio"
        tabindex="0"
      >
        <input
          class="el-radio-button__orig-radio"
          tabindex="-1"
          type="radio"
          value="all"
        />
        <span
          class="el-radio-button__inner"
          style="background-color: rgb(102, 212, 235); border-color: #66d4eb; box-shadow: -1px 0 0 0 #66D4EB; color: rgb(255, 255, 255);"
        >
          全部
          <!---->
        </span>
      </label>
       
      <label
        class="el-radio-button el-radio-button--small"
        role="radio"
        tabindex="-1"
      >
        <input
          class="el-radio-button__orig-radio"
          tabindex="-1"
          type="radio"
          value="todo"
        />
        <span
          class="el-radio-button__inner"
        >
          待完成
          <!---->
        </span>
      </label>
       
      <label
        class="el-radio-button el-radio-button--small"
        role="radio"
        tabindex="-1"
      >
        <input
          class="el-radio-button__orig-radio"
          tabindex="-1"
          type="radio"
          value="done"
        />
        <span
          class="el-radio-button__inner"
        >
          已完成
          <!---->
        </span>
      </label>
    </div>
  </div>
   
  <ul />
   
  <p
    class="encourage"
  >
    空的日程列表
  </p>
   
  <div
    class="pagination-container"
  >
    <!---->
  </div>
</div>
`;
